It enables the new multipart transfer manager inside track downloader.

fressi-elastic opened the pull request:
It allows to use the new multipart transfer manager for downloading track files.

## Configuration documentation.
- It adds the new track section

## `Adapter` interface.
- It renames adapter methods parameters: `head` -> `want`.

## `Client` class.
- Cache TTL values are now configurable (60. seconds by default).
- It uses the  new `CachedHead` `class to store cached heads.
- `Client.head` method raises `CachedHeadError` to wrap errors that have been previously cached.
- `Client.resolve` method only logs errors that hasn't been obtaining from the cache (I.E. when capturing a `CachedHeadError`).
- `Client.get` refactored to properly handle the `resolved` head filtering.

## Config
- The new `StorageConfig` class is now used to pack values parsed from the `esrally.condig.Config` object and pass values through objects of the storage package. This is nice to reduce the effort of keeping distinction between user configured values and actual object variables. It also make easier to centralize configuration parsing and validation to one single place.

## Track loader
- It add support for the new transfer manager as optional downloading engine by a `track.downloader.multipart_enabled` configuration flag.
- It lazily sets up the transfer manager the first time it requires for it so that its thread pool is created only in the actor process where it is required.
- It uses a global logger instance for logging messages. It relies on the lazy params formatting for skipping unnecessary formatting.

## Threads utils
- Check `interval` value when initializing `ContinuousTimer` class. The value is now public.

## Testing
- It updates unit tests after above changes.


**Note** this requited the following PRs to be merged first:
- [x] #1960
- [x] #1964
- [x] #1975

fressi-elastic committed:
Update storage tools with last changes from `storage` branch.

fressi-elastic committed:
Add `storage.max_workers` otpion to `types` module.

fressi-elastic committed:
It adds TrasferManager class.

fressi-elastic committed:
Add test case for TransferManager.get method.

fressi-elastic committed:
It adds missing options keys to `èsrally.types` module.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.manager

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.transfer

fressi-elastic committed:
Add documentation for `storage.max_workers` configuration key.

fressi-elastic committed:
Merge branch 'storage.transfer' into storage.manager

fressi-elastic committed:
Add documentation for new `storage.*` configuration keys.

fressi-elastic committed:
It enables the new multipart tranfer manager inside track downloader.

fressi-elastic committed:
Update configuration documentation.

fressi-elastic committed:
Use legacy downloader in case the transfer manager fails.

fressi-elastic committed:
Enable the multipart downloader to test in in CI

fressi-elastic committed:
Writing tests for Downloader class

fressi-elastic committed:
Improve loader and test using transfer manager.

fressi-elastic committed:
Backport changes from track loader integration branch.

fressi-elastic committed:
Merge branch 'storage.transfer' into storage.manager

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Remove unused adapter methods.

fressi-elastic committed:
Merge branch 'storage.transfer' into storage.manager

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
WIP making Downloader serializable.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.transfer

fressi-elastic committed:
Skip 'url' field and use 'all()' function in 'Head.check' method.

fressi-elastic committed:
Skip 'url' field and use 'all()' function in 'Head.check' method.

fressi-elastic committed:
Merge branch 'storage.transfer' of github.com:fressi-elastic/rally into storage.transfer

fressi-elastic committed:
Remove unused property

fressi-elastic committed:
Add a comment to remind to implement Data header in the http adapter.

fressi-elastic committed:
It Raises an NotImplementedError for unsupported multi-range feature in HTTP adapter.

fressi-elastic committed:
Adapter.match_url now returns a bool once again.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.transfer

fressi-elastic committed:
Merge branch 'storage.transfer' into storage.manager

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Remove unnecessary match_url method overriding.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.transfer

fressi-elastic committed:
Merge branch 'storage.transfer' into storage.manager

fressi-elastic committed:
It adds support to AWS/S3 client adapter.

fressi-elastic committed:
Testing the new AWS/S3 client adapter.

fressi-elastic committed:
Testing the new AWS/S3 client adapter.

fressi-elastic committed:
Remove Head.create method.

fressi-elastic committed:
Update HTTPAdapter to make easier to reuse its head to headers methods.

fressi-elastic committed:
It adds tests for AWS/S3 client adapter.

fressi-elastic committed:
Fix S3Adapter.put object

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.manager

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Downloader class now uses a global storage manager instance.

fressi-elastic committed:
Backport changes from track.loader branch.

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Use global LOG reference instead of self.logger.

fressi-elastic committed:
Use `Self` as return type for `Adapter.from_config` methods.

fressi-elastic committed:
Use `Self` as return type for `Adapter.from_config` methods.

fressi-elastic committed:
Use `Self` as return type for `Adapter.from_config` methods.

fressi-elastic committed:
Add comment to explain client limitation.

fressi-elastic committed:
Add missing typing for boto3 library.

fressi-elastic committed:
Add missing typing for boto3 library.

fressi-elastic committed:
Import `Self` from typing_extensions

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into storage.manager

fressi-elastic committed:
Remove list and put methods from storage Adapter interface

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Update documentation after introducing S3 adapter.

fressi-elastic committed:
Merge branch 'storage.manager' into track.loader

fressi-elastic committed:
Fix configuration documentation.

fressi-elastic committed:
Write logs lines for when using transfer manager a single transfer does start and end.

fressi-elastic committed:
Force using transfer manager

fressi-elastic committed:
Revert "Force using transfer manager"

This reverts commit f7521550262f27881699cb23de4cd45c4cf76824.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into track.loader

fressi-elastic committed:
Get improvements from storage branch.

gbanasiak reviewed (commented):
I did a code pass and it looks good to me. I'd like to spend some time testing it, but will be able to allocate some time in a week due to PTO. Perhaps someone else can do some tests in the meantime.

Can we add GCS (`gs://`) support before turning on new downloader by default? Or, alternatively, can we default to the old method and merge as is?

fressi-elastic committed:
Disable the use of the new transfer manager by default.

gbanasiak commented on esrally/storage/_client.py:
```diff
@@ -87,39 +122,41 @@ def adapters(self):
+        return head
 
-    def resolve(self, url: str, check: Head | None, ttl: float = 60.0) -> Iterator[Head]:
+    def resolve(self, url: str, want: Head | None, ttl: float = 60.0) -> Iterator[Head]:
```
I meant the value of `60.0`.

gbanasiak reviewed (changes requested):
I'm getting the following error when testing on MacOS Sequoia 15.6:

```
% esrally race --track=geonames --challenge=append-no-conflicts-index-only --distribution-version=8.15.5 --enable-assertions --kill-running-processes --on-error="abort" --test-mode

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [aece7a14-e017-4a71-aad9-fbef5dbe13a9]
[INFO] Preparing for race ...
objc[12835]: +[NSNumber initialize] may have been in progress in another thread when fork() was called.
objc[12835]: +[NSNumber initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

My `rally.ini` includes:
```
[track]
track.downloader.multipart_enabled = true
```

I've enabled Thespian debug logs:
```
export THESPLOG_FILE=/Users/grzegorz/.rally/logs/actor-system-internal.log
export THESPLOG_FILE_MAXSIZE=10737418240
export THESPLOG_THRESHOLD=DEBUG
```

The process that crashes is `NodeMechanicActor` (here `p12835`):
```
% grep "Starting Actor" actor-system-internal.log
2025-08-11 10:38:36.399433 p12741 I    Starting Actor esrally.racecontrol.BenchmarkActor at ActorAddr-(T|:50296) (parent ActorAddr-(T|:1900), admin ActorAddr-(T|:1900), srcHash None)
2025-08-11 10:38:37.202126 p12816 I    Starting Actor esrally.mechanic.mechanic.MechanicActor at ActorAddr-(T|:50317) (parent ActorAddr-(T|:50296), admin ActorAddr-(T|:1900), srcHash None)
2025-08-11 10:38:37.946286 p12834 I    Starting Actor esrally.mechanic.mechanic.Dispatcher at ActorAddr-(T|:50333) (parent ActorAddr-(T|:50317), admin ActorAddr-(T|:1900), srcHash None)
2025-08-11 10:38:37.964967 p12835 I    Starting Actor esrally.mechanic.mechanic.NodeMechanicActor at ActorAddr-(T|:50335) (parent ActorAddr-(T|:50333), admin ActorAddr-(T|:1900), srcHash None)
```

The last logs sent by this actor are:
```
2025-08-11 10:38:38.504339 p12835 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:50290)-pending-ExpiresIn_0:04:59.999974-<class 'logging.LogRecord'>-<LogRecord: esrally.mechanic.supplier, 20, /Users/grzegorz/src/rally/esrally/mechanic/supplier.py, 596, "Resolved download URL [https://artifacts.elas...-quit_0:04:59.999967)
2025-08-11 10:38:38.504780 p12835 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:50290)-pending-ExpiresIn_0:04:59.999984-<class 'logging.LogRecord'>-<LogRecord: esrally.mechanic.supplier, 20, /Users/grzegorz/src/rally/esrally/mechanic/supplier.py, 611, "Skipping download for version [8.15.5]. Found...-quit_0:04:59.999980)
2025-08-11 10:38:48.083407 p12834 dbg  ACTOR got ReceiveEnvelope(from: ActorAddr-(T|:50333), <class 'thespian.system.messages.multiproc.ChildMayHaveDied'> msg: <thespian.system.messages.multiproc.ChildMayHaveDied object at 0x105749fa0>)
```

After disabling `track.downloader.multipart_enabled` it works fine.

fressi-elastic committed:
Send storage configuration toghether with the Downloader object.

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into actor

fressi-elastic committed:
Update test_types and launcher.py

fressi-elastic committed:
Revert changes of mechanic launcher.

fressi-elastic committed:
Ask Thespian to spawn actors subprocess instead of using fork

The main reason for this is that to be able to use threads in
subprocesses we have to prevemt using any kind of fork function.

fressi-elastic committed:
Correct actor initialization procedure.

fressi-elastic committed:
revert changes to rally.py

fressi-elastic committed:
Revert unrelated changes.

fressi-elastic committed:
Revert unrelated changes.

fressi-elastic committed:
Revert unrelated changes.

fressi-elastic committed:
Revert unrelated changes.

fressi-elastic committed:
Merge branch 'actor' of github.com:fressi-elastic/rally into actor

fressi-elastic committed:
Add test case for actor.bootstrap_actor_system

fressi-elastic committed:
Add test case for actor.actor_system_already_running

fressi-elastic committed:
Test setting process_startup_method when bootstrapping actor system.

fressi-elastic committed:
It allows configuring actor process startup method.

fressi-elastic committed:
Test behaviour when joining actor system offline.

fressi-elastic committed:
actor_system_already_running returns `None` for unsupported system base.

fressi-elastic committed:
It joins the remote actor system with the requested IP and port.

fressi-elastic committed:
It wraps actor_system_already_running to keep the old behavior.

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Fix actor_test.py

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Mock log.load_configuration in test_actor.py

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Fix manager_test.py

fressi-elastic committed:
Update configuration documentation.

fressi-elastic committed:
Update configuration documentation.

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Fix/refactor bootstrap_actor_system function.

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Remove wrappers for actor_system_already_running method.

fressi-elastic committed:
Fix actor_test.py

fressi-elastic committed:
Merge branch 'actor' into track.loader

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into track.loader

fressi-elastic committed:
Revert unnecessary changes from rally.py

fressi-elastic committed:
Revert unnecessary changes in types_test.py

fressi-elastic committed:
Simplyfy trasfer manager interface.

fressi-elastic committed:
Simplyfy trasfer manager interface.

fressi-elastic committed:
Make default client TTLs configurable.

fressi-elastic commented:
@gbanasiak This is ready for review. As follow up I would like to add support for a multiprocessing pool to be able replacing the thread pool and work nicely with actors `fork` process startup method that is still the default. Even if on Linux is not mandatory using `spawn`, we could still experience problems using threads (even if we didn't saw any yet).

gbanasiak commented:
@fressi-elastic The `spawn` process creation method does not work for me. In the meantime I have updated MacOS to 15.6.1, not sure if that's relevant. I'm getting the same symptoms with this PR as well as master branch, so that's after https://github.com/elastic/rally/pull/1975.

```
[actor]
actor.process.startup.method = spawn
```

```
% esrally --version
esrally 2.12.0.dev0 (git revision: 81abfc385aa27a785c43dfb98fddc58dddcefc9c)

% esrally race --track=geonames --challenge=append-no-conflicts-index-only --distribution-version=8.15.5 --enable-assertions --kill-running-processes --on-error="abort" --test-mode

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [68b2002a-858e-44ca-ae25-ceaacc0b86be]
[ERROR] Cannot race. This race ended with a fatal crash.
```

`rally.log` (UTC):
```
2025-08-21 19:56:04,686 -not-actor-/PID:97283 esrally.rally DEBUG Command line: /Users/grzegorz/src/rally/.venv/bin/esrally race --track=geonames --challenge=append-no-conflicts-index-only --distribution-version=8.15.5 --enable-assertions --kill-running-processes --on-error=abort --test-mode
2025-08-21 19:56:04,719 -not-actor-/PID:97283 esrally.rally INFO OS [uname_result(system='Darwin', node='pc-11.home', release='24.6.0', version='Darwin Kernel Version 24.6.0: Mon Jul 14 11:28:30 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T6030', machine='arm64')]
2025-08-21 19:56:04,719 -not-actor-/PID:97283 esrally.rally INFO Python [namespace(name='cpython', cache_tag='cpython-312', version=sys.version_info(major=3, minor=12, micro=2, releaselevel='final', serial=0), hexversion=51118832, _multiarch='darwin')]
2025-08-21 19:56:04,742 -not-actor-/PID:97283 esrally.rally INFO Rally version [2.12.0.dev0 (git revision: 81abfc385aa27a785c43dfb98fddc58dddcefc9c)]
2025-08-21 19:56:04,742 -not-actor-/PID:97283 esrally.rally DEBUG Command line arguments: Namespace(subcommand='race', cluster_name='rally-benchmark', distribution_version='8.15.5', team_path=None, team_repository='default', team_revision=None, race_id='68b2002a-858e-44ca-ae25-ceaacc0b86be', pipeline='', revision='current', track_repository='default', track_path=None, track_revision=None, track='geonames', track_params='', challenge='append-no-conflicts-index-only', car='defaults', car_params='', runtime_jdk=None, elasticsearch_plugins='', plugin_params='', target_hosts='', load_driver_hosts='localhost', client_options='timeout:60', on_error='abort', telemetry='', telemetry_params='', distribution_repository='release', include_tasks=None, exclude_tasks=None, user_tags='', report_format='markdown', report_numbers_align='decimal', show_in_report='available', report_file='', preserve_install='false', test_mode=True, enable_driver_profiling=False, enable_assertions=True, kill_running_processes=True, source_build_method='default', effective_start_date=None, skip_rest_api_check=False, configuration_name=None, quiet=False, offline=False)
2025-08-21 19:56:04,742 -not-actor-/PID:97283 esrally.utils.net INFO Connecting directly to the Internet (no proxy support) for [all_proxy].
2025-08-21 19:56:04,742 -not-actor-/PID:97283 esrally.utils.net INFO Connecting directly to the Internet (no proxy support) for [all_proxy].
2025-08-21 19:56:04,742 -not-actor-/PID:97283 esrally.rally INFO Cleaning track dependency directory [/Users/grzegorz/.rally/libs]...
2025-08-21 19:56:04,743 -not-actor-/PID:97283 esrally.rally INFO Killing running Rally processes
2025-08-21 19:56:04,803 -not-actor-/PID:97283 esrally.rally INFO Actor system already running locally? [False]
2025-08-21 19:56:05,290 -not-actor-/PID:97283 esrally.racecontrol INFO Race id is [68b2002a-858e-44ca-ae25-ceaacc0b86be]
2025-08-21 19:56:05,290 -not-actor-/PID:97283 esrally.racecontrol INFO User specified no pipeline. Automatically derived pipeline [from-distribution].
2025-08-21 19:56:05,291 -not-actor-/PID:97283 esrally.racecontrol INFO Setting default host to [127.0.0.1:39200]
2025-08-21 19:56:58,298 -not-actor-/PID:97283 esrally.rally INFO Attempting to shutdown internal actor system.
2025-08-21 19:56:58,311 -not-actor-/PID:97345 root INFO ActorSystem Logging Shutdown
2025-08-21 19:56:58,336 -not-actor-/PID:97283 esrally.rally INFO Actor system is still running. Waiting...
2025-08-21 19:56:59,342 -not-actor-/PID:97283 esrally.rally INFO Shutdown completed.
2025-08-21 19:56:59,343 -not-actor-/PID:97283 esrally.rally ERROR Cannot run subcommand [race].
Traceback (most recent call last):
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 405, in run
    pipeline(cfg)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 74, in __call__
    self.target(cfg)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 337, in from_distribution
    return race(cfg, distribution=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 292, in race
    benchmark_actor = actor_system.createActor(BenchmarkActor, targetActorRequirements={"coordinator": True})
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/.venv/lib/python3.12/site-packages/thespian/actors.py", line 704, in createActor
    return self._systemBase.newPrimaryActor(actorClass,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/.venv/lib/python3.12/site-packages/thespian/system/systemBase.py", line 214, in newPrimaryActor
    raise ActorSystemRequestTimeout(
thespian.actors.ActorSystemRequestTimeout: No response received to PendingActor request to Admin at ActorAddr-(T|:1900) from ActorAddr-(T|:63751)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/grzegorz/src/rally/esrally/rally.py", line 1228, in dispatch_sub_command
    race(cfg, args.kill_running_processes)
  File "/Users/grzegorz/src/rally/esrally/rally.py", line 964, in race
    with_actor_system(racecontrol.run, cfg)
  File "/Users/grzegorz/src/rally/esrally/rally.py", line 1002, in with_actor_system
    runnable(cfg)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 414, in run
    raise exceptions.RallyError("This race ended with a fatal crash.").with_traceback(tb)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 405, in run
    pipeline(cfg)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 74, in __call__
    self.target(cfg)
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 337, in from_distribution
    return race(cfg, distribution=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/esrally/racecontrol.py", line 292, in race
    benchmark_actor = actor_system.createActor(BenchmarkActor, targetActorRequirements={"coordinator": True})
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/.venv/lib/python3.12/site-packages/thespian/actors.py", line 704, in createActor
    return self._systemBase.newPrimaryActor(actorClass,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/grzegorz/src/rally/.venv/lib/python3.12/site-packages/thespian/system/systemBase.py", line 214, in newPrimaryActor
    raise ActorSystemRequestTimeout(
esrally.exceptions.RallyError: This race ended with a fatal crash.
```

Thespian debug logs (UTC+2):
```
2025-08-21 21:56:05.058477 p97344 dbg  ++++ Starting Admin from /Users/grzegorz/src/rally/.venv/lib/python3.12/site-packages/thespian/__init__.py
2025-08-21 21:56:05.059110 p97344 I    ++++ Admin started @ ActorAddr-(T|:1900) / gen (3, 10) / capabilities {'coordinator': True, 'ip': '127.0.0.1', 'Convention Address.IPv4': '127.0.0.1', 'Process Startup Method': 'spawn', 'Thespian ActorSystem Name': 'multiprocTCPBase', 'Thespian ActorSystem Version': 2, 'Thespian Watch Supported': True, 'Python Version': (3, 12, 2, 'final', 0), 'Thespian Generation': (3, 10), 'Thespian Version': '1755806164818'}
2025-08-21 21:56:05.284762 p97344 dbg  Admin of ReceiveEnvelope(from: ActorAddr-(T|:63754), <class 'thespian.system.messages.multiproc.LoggerConnected'> msg: <thespian.system.messages.multiproc.LoggerConnected object at 0x104a26360>)
2025-08-21 21:56:05.285550 p97344 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:63751)-pending-ExpiresIn_0:04:59.999930-<class 'thespian.system.messages.multiproc.EndpointConnected'>-<thespian.system.messages.multiproc.EndpointConnected object at 0x1048d6030>-quit_0:04:59.999924)
2025-08-21 21:56:05.288314 p97283 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:1900)-pending-ExpiresIn_0:04:59.999904-<class 'thespian.system.messages.admin.QueryExists'>-<thespian.system.messages.admin.QueryExists object at 0x104d25820>-quit_0:04:59.999898)
2025-08-21 21:56:05.288869 p97344 dbg  Admin of ReceiveEnvelope(from: ActorAddr-(T|:63751), <class 'thespian.system.messages.admin.QueryExists'> msg: <thespian.system.messages.admin.QueryExists object at 0x104a26630>)
2025-08-21 21:56:05.289649 p97344 dbg  Attempting intent TransportIntent(ActorAddr-(T|:63751)-pending-ExpiresIn_0:04:59.999645-<class 'thespian.system.messages.admin.QueryAck'>-<thespian.system.messages.admin.QueryAck object at 0x104a26360>-quit_0:04:59.999641)
2025-08-21 21:56:05.290074 p97344 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:63751)-pending-ExpiresIn_0:04:59.999216-<class 'thespian.system.messages.admin.QueryAck'>-<thespian.system.messages.admin.QueryAck object at 0x104a26360>-quit_0:04:59.999213)
2025-08-21 21:56:05.291628 p97283 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:1900)-pending-ExpiresIn_0:04:59.999970-<class 'thespian.system.messages.admin.PendingActor'>-PendingActor#1_of_None-quit_0:04:59.999965)
2025-08-21 21:56:05.293364 p97344 dbg  Admin of ReceiveEnvelope(from: ActorAddr-(T|:63758), <class 'thespian.system.messages.admin.PendingActor'> msg: PendingActor#1_of_None)
2025-08-21 21:56:05.293770 p97344 I    Pending Actor request received for esrally.racecontrol.BenchmarkActor reqs {'coordinator': True} from ActorAddr-(T|:63758)
2025-08-21 21:56:05.526135 p97344 ERR  child ActorAddr-LocalAddr.0 is dead but cannot translate address to properly handle it
2025-08-21 21:56:58.300640 p97283 I    ActorSystem shutdown requested.
2025-08-21 21:56:58.301852 p97283 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:1900)-pending-ExpiresIn_0:04:59.999750-<class 'thespian.system.messages.admin.SystemShutdown'>-<thespian.system.messages.admin.SystemShutdown object at 0x104c3e6c0>-quit_0:04:59.999695)
2025-08-21 21:56:58.304119 p97344 dbg  Admin of ReceiveEnvelope(from: ActorAddr-(T|:63751), <class 'thespian.system.messages.admin.SystemShutdown'> msg: <thespian.system.messages.admin.SystemShutdown object at 0x104a25910>)
2025-08-21 21:56:58.305527 p97344 dbg  ---- shutdown initiated by ActorAddr-(T|:63751)
2025-08-21 21:56:58.307084 p97344 dbg  Attempting intent TransportIntent(ActorAddr-(T|:63758)-pending-ExpiresIn_0:04:59.999369-<class 'thespian.system.messages.admin.PendingActorResponse'>-PendingActorResponse(for None inst# 1) errCode 3585 actual None-quit_0:04:59.999352)
2025-08-21 21:56:58.308083 p97344 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:63758)-pending-ExpiresIn_0:04:59.998387-<class 'thespian.system.messages.admin.PendingActorResponse'>-PendingActorResponse(for None inst# 1) errCode 3585 actual None-quit_0:04:59.998376)
2025-08-21 21:56:58.309558 p97344 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:63754)-pending-ExpiresIn_0:04:59.999882-<class 'thespian.system.logdirector.LoggerExitRequest'>-<thespian.system.logdirector.LoggerExitRequest object at 0x1032f78f0>-quit_0:04:59.999872)
2025-08-21 21:56:58.333822 p97344 dbg  Attempting intent TransportIntent(ActorAddr-(T|:63751)-pending-ExpiresIn_0:04:59.999234-<class 'thespian.system.messages.admin.SystemShutdownCompleted'>-<thespian.system.messages.admin.SystemShutdownCompleted object at 0x104a262d0>-quit_0:04:59.999211)
2025-08-21 21:56:58.334730 p97344 dbg  actualTransmit of TransportIntent(ActorAddr-(T|:63751)-pending-ExpiresIn_0:04:59.998198-<class 'thespian.system.messages.admin.SystemShutdownCompleted'>-<thespian.system.messages.admin.SystemShutdownCompleted object at 0x104a262d0>-quit_0:04:59.998191)
2025-08-21 21:56:58.335458 p97344 I    ---- shutdown completed
2025-08-21 21:56:58.335891 p97283 I    ActorSystem shutdown complete.
2025-08-21 21:56:58.336435 p97344 dbg  Admin time to die
```

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into track.loader

fressi-elastic committed:
Backport changes from other PRs

fressi-elastic committed:
Fix transfer manager.

fressi-elastic committed:
Fix make install to restore the old behavior.

fressi-elastic committed:
Rename function after last day refactoring.

fressi-elastic committed:
Remove files added by mistake.

fressi-elastic committed:
Remove unnecessary lines.

fressi-elastic committed:
Update configuration documentation.

fressi-elastic committed:
Fix Makefile -> make docs

fressi-elastic committed:
It adds http and s3 adapters to the main package module.

fressi-elastic committed:
Update configuration document style.

fressi-elastic committed:
Remove file added by mistake.

fressi-elastic committed:
Remove [track] section platform issue mention from configuration documentation.

elasticmachine committed:
Merge branch 'master' into track.loader

gbanasiak commented on docs/configuration.rst:
```diff
@@ -294,6 +298,20 @@ Configuration options:
+This section specifies how tracks corpora files has to be fetched. Available options are:
+
+* ``track.downloader.multipart_enabled`` if `true`, it will enable the use the new multipart `esrally.storage` package for
+  downloading corpora files. For more configuration options please have a look to the `[storage]` configuration section.
```
The link to storage section doesn't work.

gbanasiak commented on esrally/config.py:
```diff
@@ -124,22 +131,66 @@ def auto_load_local_config(base_config, additional_sections=None, config_file_cl
 
-class Config:
-    EARLIEST_SUPPORTED_VERSION = 17
+CONFIG = contextvars.ContextVar[typing.Optional[types.Config]](f"{__name__}.config", default=None)
```
What is the problem this contextvar addresses?

fressi-elastic replied:
I would like to be able to pass a shared configuration object from the very first parent thread (produced after parsing the cmd line), and from actor to actor so we do have global settings always available in the app without having to care to pass them. We should be able to deploy the whole app from it in a deterministic and reproducible way. It would make all the parts far easier to test being a functions of it. Something similar of how I implemented the storage part.

gbanasiak replied:
I don't understand how contextvars facilitate sharing. Each thread has it's own context.

Example:
```
import concurrent.futures
import contextvars

VAR: contextvars.ContextVar[str | None] = contextvars.ContextVar('var', default=None)
VAR.set('parent')
print(f"parent thread: {VAR.get()}") # prints 'parent'

def worker() -> None:
    print(f"child thread: {VAR.get()}") # prints 'None'
    VAR.set('child')
    print(f"child thread: {VAR.get()}") # prints 'child'

with concurrent.futures.ThreadPoolExecutor(max_workers=1) as executor:
    future = executor.submit(worker)
    future.result()

print(f"parent thread: {VAR.get()}") # prints 'parent'
```

gbanasiak commented on esrally/config.py:
```diff
@@ -124,22 +131,66 @@ def auto_load_local_config(base_config, additional_sections=None, config_file_cl
+        if isinstance(cfg, cls):
+            return cfg
+        if isinstance(cfg, types.Config):
+            return cls(opts_from=cfg)
```
Where is `opts_from` defined?

gbanasiak commented on esrally/storage/_manager.py:
```diff
@@ -15,128 +15,109 @@
+            except Exception as ex:
+                LOG.error("error closing transfer: %s, %s", tr.url, ex)
         self._monitor_timer.cancel()
+        LOG.info("Project manager shut down.")
```
```suggestion
        LOG.info("Transfer manager shut down.")
```

BTW, I do not see this log in `rally.log`. It does appear in Thespian logs, but apparently logger actor gets shutdown before it manages to process that?

gbanasiak reviewed (commented):
I left a few comments.

fressi-elastic commented on Makefile:
```diff
@@ -113,6 +113,7 @@ install_pytest_rally_plugin: venv
 
 # Old legacy alias goals
 install: venv
+	uv run esrally --version
```
I agree. I should do it as suggested. It was a quick hack indeed 😁

fressi-elastic commented on esrally/storage/_client.py:
```diff
@@ -22,108 +22,134 @@
+            try:
+                raise CachedHeadError("Cached exception") from error
+            except CachedHeadError as ex:
+                self.error = ex
```
It would not have the stack trace set. Probably I can do it without raising copying it from the wrapped exception. I'll look.

gbanasiak replied:
I'm also fine with a comment. I just paused reading this.

fressi-elastic committed:
Recover all known loggers by copying __dict__ reference from new to old loggers.

gbanasiak commented:
@fressi-elastic Regarding IT failure, `boto3` became mandatory due to `S3Adapter`. Maybe keep it optional for BWC for now and reduce default adapters to `HTTPAdapter` ?

fressi-elastic committed:
Some further iteration with all loggers recovered and actor using the logger of its class definition module.

fressi-elastic committed:
Refactor typing annotations and improve code readability.

fressi-elastic committed:
Merge branch 'log' into track.loader

fressi-elastic committed:
Merge branch 'track.loader' of github.com:fressi-elastic/rally into track.loader

fressi-elastic committed:
Use `uv sync` in `make install` goal.

fressi-elastic committed:
Fix link to storage section in configuration.rst

fressi-elastic committed:
Fix docstring in _adapter.py

fressi-elastic committed:
Explain how an why the CachedHeadException is being created.

fressi-elastic committed:
Update esrally/storage/_manager.py

Co-authored-by: Grzegorz Banasiak <grzegorz.banasiak@elastic.co>

fressi-elastic committed:
Remove unused attribute.

fressi-elastic committed:
Remove from __future__ import annotations.

fressi-elastic committed:
Update make *docs rules to use `uv run`.

fressi-elastic commented on pyproject.toml:
```diff
@@ -130,6 +130,8 @@ develop = [
     "types-requests<2.32.0",
     "types-jsonschema==3.2.0",
+    # Python dead library removed in version 3.13 and used to build Rally docs.
+    "standard-imghdr==3.13.0",
```
It doesn't work on my case. I retried it. `make docs` does require that module with Python 3.13.

fressi-elastic committed:
Create local directory only before starting download to the final destination.

fressi-elastic commented:
> > @fressi-elastic Regarding IT failure, `boto3` became mandatory due to `S3Adapter`. Maybe keep it optional for BWC for now and reduce default adapters to `HTTPAdapter` ?
> 
> S3Adapter is only required to access buckets using credentials. For public buckets we should be safe using HTTPAdapter even with S3. The difference is in the URL as specified in the documentation. rclone should be kind enoght to set up permissions for uploading files with public read access on S3 buckets, so if we want to make publicly available some mirrors then se can omit boto3 installation for general availability of those mirros.

BTW when boto3 is not available, storage client should fail registering S3Adapter and simply skip it. On such a case `s3://` url will not be supported, but the same bucket could be reached using the appropriate `https://` URL instead.

fressi-elastic committed:
Some other additional cleanup for readibility to help identify integration test failure cause.

fressi-elastic committed:
It updates logging configuration after actor creation.

fressi-elastic committed:
Silence eslastic.log logger.

fressi-elastic committed:
Backport changes from branch track.loader

fressi-elastic committed:
Remove unnecessary checks.

fressi-elastic committed:
Merge branch 'log' into track.loader

fressi-elastic commented on esrally/storage/_manager.py:
```diff
@@ -194,36 +175,37 @@ def _update_transfers(self) -> None:
 
-_LOCK = threading.Lock()
-_MANAGER: TransferManager | None = None
+_MANAGER = contextvars.ContextVar[TransferManager | None](f"{__name__}.transfer_manager", default=None)
```
Yes at this moment in time there could be one transfer manager for every thread. The actual idea should be having one transfer manager for every host. In the middle of both there is the option of having a transfer for process, that it still desired and could be archived by copying the context from the main thread after transfer manager initialization. Given the fact there is only one worker thread per worker process, for simplicity and time constrains I would leave this as ti is now with the purpose or going forward of having a transfer manager process (actor) later in a later iteration. @gbanasiak Please give a look to PR #1988 before coming back to this one.

fressi-elastic committed:
Remove unrelated changes.

fressi-elastic committed:
Merge branch 'log' into track.loader

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into track.loader

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into log

fressi-elastic committed:
Merge branch 'log' into track.loader

fressi-elastic committed:
Merge branch 'master' of github.com:elastic/rally into track.loader

fressi-elastic committed:
Make S3 adapter optional (once again).

fressi-elastic committed:
Cache missing adapter error to avoid spamming the logging with errors when `boto3` package is not installed.

fressi-elastic committed:
Rename http package and make it used from its final implementation module.

gbanasiak reviewed (approved):
LGTM. Thank you for your patience.

fressi-elastic committed:
Correct configuration documentation.

fressi-elastic committed:
Relax test_execute_schedule_throughput_throttled test case requirements.
